torch/_functorch/autograd_function.py

```diff
@@ -0,0 +1,193 @@
+            # to have the same object identity as the input.
+            # Mode-only functorch will greatly simplify this logic.
+            return wrap_outputs_maintaining_identity(
+                output,
+                unwrapped_operands,
+                operands,
+                level)
+
+        @staticmethod
+        def setup_context(ctx, outputs, *operands):
+            return autograd_function.setup_context(ctx, outputs, *operands)
+
+        @staticmethod
+        def backward(ctx, *grads):
+            result = autograd_function.backward(ctx, *grads)
+            return result
+
+    with enable_autograd_function():
+        flat_out = Generated.apply(*operands)
+    return flat_out
+
+
+def wrap_outputs_maintaining_identity(outputs, unwrapped_inputs, orig_inputs, level):
+    flat_output, _ = pytree.tree_flatten(outputs)
+    flat_unwrapped_inputs, _ = pytree.tree_flatten(unwrapped_inputs)
+    flat_orig_inputs, _ = pytree.tree_flatten(orig_inputs)
+
+    unwrapped_input_to_orig_input = {
+        id(unwrapped): orig
+        for unwrapped, orig in zip(flat_unwrapped_inputs, flat_orig_inputs)
+    }
+
+    flat_outputs, spec = pytree.tree_flatten(outputs)
+    result = []
+
+    for output in flat_outputs:
+        if not isinstance(output, torch.Tensor):
+            result.append(output)
+            continue
+        if id(output) in unwrapped_input_to_orig_input:
+            result.append(unwrapped_input_to_orig_input[id(output)])
+            continue
+        result.append(_wrap_for_grad(output, level))
+
+    return pytree.tree_unflatten(result, spec)
+
+
+# NOTE: [functorch vjp and autograd interaction]
+# There's an edge case with the functorch vjp and autograd interaction
+# that will eventually be fixed by mode-only functorch.
+# The TL;DR is that there's no way to unwrap a dead GradTensorWrapper,
+# so we (the framework) need to do it manually. Regular PyTorch operators
+# automatically do so this is consisent.
+#
+# class MyExp(torch.autograd.Function):
+#     @staticmethod
+#     def forward(x):
+#         return x.exp()
+#
+#     @staticmethod
+#     def setup_context(ctx, outputs, x):
+#         y = outputs
+#         ctx.save_for_backward(y)
+#
+#     @staticmethod
+#     def backward(gy):
+#         y, = ctx.saved_tensors()
+#         return MyMul.apply(gy, y)
+#
+# x = torch.randn([], requires_grad=True)
+# gy = torch.randn([], requires_grad=True)
+# _, vjp_fn = vjp(MySin.apply, x)
+# result = vjp_fn(gy)
+#
+# MyMul is an autograd.Function that is not shown here.
+# It saves a `y` for backward (since gy requires grad).
+#
+# in vjp_fn(gy), we get:
+# > MyMul.apply(gy, GradTensorWrapper(y, level=dead))
+# Because the y that is saved for backward by MyExp is a GradTensorWrapper
+# but is now dead since we are outside the vjp context.
+#
+# PyTorch dispatcher operations, upon seeing a dead GradTensorWrapper,
+# will automatically unwrap the GradTensorWrapper when applied.
```

Comment:
Yes, this refers to the dead_tensor_wrapper_fallback.

Let's assume an input to the autograd.Function is a dead tensor wrapper.
It needs to be handled at this level because otherwise ctx.save_for_backward, on saving an input for backward, will save a dead tensor wrapper. This will mess up the autograd graph, because it'll look like the graph is disconnected from the underlying Tensor.

Contrast this to what happens in an aten pytorch operator. In an aten pytorch operator, saving Tensors for backward happens AFTER a tensor has dispatched on the autograd key. dead_tensor_wrapper_fallback runs before the Autograd key, so the VariableType kernel ends up saving the underlying Tensor instead of a dead tensor wrapper.

There are some other problems that we can run into if we don't unwrap dead tensor wrapper here. For example, it is incorrect to directly pass a dead tensor wrapper to the autograd.Function's forward pass because it may grab the .data_ptr of a dead tensor wrapper and error as a result.

> By "consistent" do you mean you want to autograd function to be consistent with normal torch operators?

Yes.

Reply:
Makes sense, thanks!